File: apps/ios/FishKit/Sources/ChatCore/Logic/ChatReducer.swift
import Foundation

public enum ChatStateReducer {
    public static func createEmptyChatState() -> ChatState { ChatState() }

    public static func apply(_ events: [ChatEvent], to state: ChatState) -> ChatState {
        events.reduce(state, reduce)
    }

    public static func reduce(_ state: ChatState, _ event: ChatEvent) -> ChatState {
        switch event {
        case .hydrateConversation(let id, let messages, let readStates):
            var conversation = conversation(in: state, id: id)
            conversation.messages = mergeHydratedMessages(
                conversation.messages,
                messages.map { normalize($0, status: .sent) }
            )
            conversation.readStates = readStates
            return setting(conversation, in: state)

        case .draftChanged(let id, let draft):
            return updating(state, id: id) { $0.composer.draft = draft }

        case .sendOptimisticMessage(let message):
            return merging(normalize(message, status: .sending), into: state)

        case .confirmSentMessage(let message, let localRequestId):
            return merging(
                normalize(strippingFailure(message), status: .sent),
                into: state,
                localRequestId: localRequestId
            )

        case .markMessageFailed(let id, let requestId, let reason):
            return markFailed(state, conversationId: id, requestId: requestId, reason: reason)

        case .mergeRemoteMessage(let message, let localRequestId):
            return merging(
                normalize(strippingFailure(message), status: .sent),
                into: state,
                localRequestId: localRequestId
            )

        case .mergeReadState(let id, let readState):
            return updating(state, id: id) {
                $0.readStates = ChatSelectors.mergeReadState($0.readStates, readState)
            }

        case .setReplyTarget(let id, let messageId):
            return updating(state, id: id) { $0.composer.replyTargetId = messageId }

        case .setEditTarget(let id, let messageId):
            return updating(state, id: id) { $0.composer.editTargetId = messageId }

        case .setRealtimeStatus(let id, let status):
            return updating(state, id: id) { $0.realtime.status = status }

        case .clearComposer(let id):
            return updating(state, id: id) { $0.composer = ChatComposerState() }

        case .hydrateWindow(let id, let messages, let reads, let hasMore, let cursor):
            var conversation = conversation(in: state, id: id)
            conversation.messages = mergeHydratedMessages(
                conversation.messages,
                messages.map { normalize($0, status: .sent) }
            )
            conversation.readStates = reads
            conversation.pagination = ChatPaginationState(
                oldestLoadedCursor: cursor,
                hasMoreOlder: hasMore
            )
            return setting(conversation, in: state)

        case .olderMessagesRequested(let id):
            return updating(state, id: id) {
                guard !$0.pagination.isLoadingOlder else { return }
                $0.pagination.isLoadingOlder = true
                $0.pagination.hasLoadError = false
            }

        case .olderPageLoaded(let id, let messages, let hasMore, let cursor):
            return updating(state, id: id) { conversation in
                conversation.messages = messages.reduce(conversation.messages) {
                    ChatSelectors.mergeChatMessage($0, normalize($1, status: .sent))
                }
                conversation.pagination = ChatPaginationState(
                    oldestLoadedCursor: cursor,
                    hasMoreOlder: hasMore
                )
            }

        case .olderPageLoadFailed(let id):
            return updating(state, id: id) {
                guard $0.pagination.isLoadingOlder else { return }
                $0.pagination.isLoadingOlder = false
                $0.pagination.hasLoadError = true
            }
        }
    }

    private static func merging(
        _ message: ChatMessageState,
        into state: ChatState,
        localRequestId: String? = nil
    ) -> ChatState {
        updating(state, id: message.conversationId) {
            $0.messages = ChatSelectors.mergeChatMessage(
                $0.messages,
                message,
                localRequestId: localRequestId
            )
        }
    }

    private static func mergeHydratedMessages(
        _ existing: [ChatMessageState],
        _ incoming: [ChatMessageState]
    ) -> [ChatMessageState] {
        let unresolved = existing.filter {
            $0.localStatus == .pending || $0.localStatus == .sending || $0.localStatus == .failed
        }
        return incoming.reduce(unresolved) {
            ChatSelectors.mergeChatMessage($0, $1)
        }.sorted(
            by: ChatSelectors.compareChatMessages
        )
    }

    private static func markFailed(
        _ state: ChatState,
        conversationId: String,
        requestId: String,
        reason: String?
    ) -> ChatState {
        updating(state, id: conversationId) { conversation in
            let failed = conversation.messages.first { $0.clientRequestId == requestId }
            guard failed?.localStatus != .sent else { return }
            let restoreDraft = conversation.composer.draft.isEmpty
            conversation.messages = conversation.messages.map { message in
                guard message.clientRequestId == requestId else { return message }
                var next = message
                next.localStatus = .failed
                next.failureReason = reason
                return next
            }
            if restoreDraft { conversation.composer.draft = failed?.body ?? "" }
        }
    }

    private static func normalize(
        _ message: ChatMessageState,
        status: LocalMessageStatus
    ) -> ChatMessageState {
        var next = message
        next.reactions = message.reactions ?? []
        next.localStatus = status
        return next
    }

    private static func strippingFailure(_ message: ChatMessageState) -> ChatMessageState {
        var next = message
        next.failureReason = nil
        return next
    }

    private static func conversation(in state: ChatState, id: String) -> ChatConversationState {
        state.conversations[id] ?? ChatConversationState(conversationId: id)
    }

    private static func updating(
        _ state: ChatState,
        id: String,
        _ update: (inout ChatConversationState) -> Void
    ) -> ChatState {
        var conversation = conversation(in: state, id: id)
        let original = conversation
        update(&conversation)
        guard conversation != original || state.conversations[id] == nil else { return state }
        return setting(conversation, in: state)
    }

    private static func setting(_ conversation: ChatConversationState, in state: ChatState) -> ChatState {
        var next = state
        next.conversations[conversation.conversationId] = conversation
        return next
    }
}

public func createEmptyChatState() -> ChatState { ChatStateReducer.createEmptyChatState() }

public func applyChatEvents(_ state: ChatState, _ events: [ChatEvent]) -> ChatState {
    ChatStateReducer.apply(events, to: state)
}

public func reduceChatState(_ state: ChatState, _ event: ChatEvent) -> ChatState {
    ChatStateReducer.reduce(state, event)
}

File: apps/android/feature/chat/src/main/kotlin/space/fishhub/android/feature/chat/state/ChatReducer.kt
package space.fishhub.android.feature.chat.state

import space.fishhub.android.data.chat.model.ChatMessage
import space.fishhub.android.data.chat.model.LocalMessageStatus

fun reduceChatState(state: ChatState, event: ChatEvent): ChatState = when (event) {
    is ChatEvent.HydrateConversation -> {
        val conversation = state.conversation(event.conversationId)
        state.withConversation(
            conversation.copy(
                messages = mergeHydratedMessages(
                    conversation.messages,
                    event.messages.map { it.normalized(LocalMessageStatus.Sent) },
                ),
                readStates = event.readStates,
            ),
        )
    }
    is ChatEvent.DraftChanged -> state.updateConversation(event.conversationId) {
        copy(composer = composer.copy(draft = event.draft))
    }
    is ChatEvent.SendOptimisticMessage -> state.mergeMessage(
        event.message.normalized(LocalMessageStatus.Sending),
    )
    is ChatEvent.ConfirmSentMessage -> state.mergeMessage(
        event.message.stripFailure().normalized(LocalMessageStatus.Sent),
        event.localRequestId ?: event.message.clientRequestId,
    )
    is ChatEvent.MarkMessageFailed -> state.markMessageFailed(event)
    is ChatEvent.MergeRemoteMessage -> state.mergeMessage(
        event.message.stripFailure().normalized(LocalMessageStatus.Sent),
        event.localRequestId ?: event.message.clientRequestId,
    )
    is ChatEvent.MergeReadState -> state.updateConversation(event.conversationId) {
        copy(readStates = mergeReadState(readStates, event.readState))
    }
    is ChatEvent.SetReplyTarget -> state.updateConversation(event.conversationId) {
        copy(composer = composer.copy(replyTargetId = event.messageId))
    }
    is ChatEvent.SetEditTarget -> state.updateConversation(event.conversationId) {
        copy(composer = composer.copy(editTargetId = event.messageId))
    }
    is ChatEvent.SetRealtimeStatus -> state.updateConversation(event.conversationId) {
        val hasConnected = realtime.hasConnected || event.status == RealtimeConnectionStatus.Connected
        if (realtime.status == event.status && realtime.hasConnected == hasConnected) this else copy(
            realtime = realtime.copy(status = event.status, hasConnected = hasConnected),
        )
    }
    is ChatEvent.ClearComposer -> state.updateConversation(event.conversationId) {
        copy(composer = ChatComposerState())
    }
    is ChatEvent.HydrateWindow -> {
        val conversation = state.conversation(event.conversationId)
        state.withConversation(
            conversation.copy(
                messages = mergeHydratedMessages(
                    conversation.messages,
                    event.messages.map { it.normalized(LocalMessageStatus.Sent) },
                ),
                readStates = event.readStates,
                pagination = ChatPaginationState(
                    oldestLoadedCursor = event.oldestCursor,
                    hasMoreOlder = event.hasMoreOlder,
                ),
            ),
        )
    }
    is ChatEvent.OlderMessagesRequested -> state.updateConversation(event.conversationId) {
        if (pagination.isLoadingOlder) this else copy(
            pagination = pagination.copy(isLoadingOlder = true, hasLoadError = false),
        )
    }
    is ChatEvent.OlderPageLoaded -> state.updateConversation(event.conversationId) {
        copy(
            messages = event.messages.fold(messages) { current, message ->
                mergeChatMessage(current, message.normalized(LocalMessageStatus.Sent))
            },
            pagination = ChatPaginationState(
                oldestLoadedCursor = event.oldestCursor,
                hasMoreOlder = event.hasMoreOlder,
            ),
        )
    }
    is ChatEvent.OlderPageLoadFailed -> state.updateConversation(event.conversationId) {
        if (!pagination.isLoadingOlder) this else copy(
            pagination = pagination.copy(isLoadingOlder = false, hasLoadError = true),
        )
    }
}

fun applyChatEvents(state: ChatState, events: List<ChatEvent>): ChatState =
    events.fold(state, ::reduceChatState)

private fun ChatState.mergeMessage(
    message: ChatMessage,
    localRequestId: String = message.clientRequestId,
): ChatState = updateConversation(message.conversationId) {
    copy(messages = mergeChatMessage(messages, message, localRequestId))
}

private fun mergeHydratedMessages(
    existing: List<ChatMessage>,
    incoming: List<ChatMessage>,
): List<ChatMessage> {
    val unresolved = existing.filter {
        it.localStatus == LocalMessageStatus.Pending ||
            it.localStatus == LocalMessageStatus.Sending ||
            it.localStatus == LocalMessageStatus.Failed
    }
    return incoming.fold(unresolved) { current, message ->
        mergeChatMessage(current, message)
    }.sortedWith(ChatMessageComparator)
}

private fun ChatState.markMessageFailed(event: ChatEvent.MarkMessageFailed): ChatState =
    updateConversation(event.conversationId) {
        val failedMessage = messages.firstOrNull { it.clientRequestId == event.clientRequestId }
        if (failedMessage?.localStatus == LocalMessageStatus.Sent) return@updateConversation this
        val restoredDraft = if (composer.draft.isEmpty()) failedMessage?.body ?: composer.draft else {
            composer.draft
        }
        copy(
            messages = messages.map { message ->
                if (message.clientRequestId == event.clientRequestId) {
                    message.copy(
                        localStatus = LocalMessageStatus.Failed,
                        failureReason = event.reason,
                    )
                } else {
                    message
                }
            },
            composer = composer.copy(draft = restoredDraft),
        )
    }

private fun ChatState.updateConversation(
    conversationId: String,
    transform: ChatConversationState.() -> ChatConversationState,
): ChatState = withConversation(conversation(conversationId).transform())

private fun ChatState.withConversation(conversation: ChatConversationState): ChatState = copy(
    conversations = conversations + (conversation.conversationId to conversation),
)

private fun ChatState.conversation(conversationId: String): ChatConversationState =
    conversations[conversationId] ?: ChatConversationState(conversationId = conversationId)

private fun ChatMessage.normalized(status: LocalMessageStatus): ChatMessage = copy(
    reactions = reactions,
    localStatus = status,
)

private fun ChatMessage.stripFailure(): ChatMessage = copy(failureReason = null)
